### PR TITLE
#148 - Database setup with seeding

### LINKED ISSUE #148: Prepare database structure
We should think about what kind of data we'd like to store in the project's database. 
 We need to review what's necessary or what can be stored as a static resource. 

 Some basic tables I can think of: 
 - Accounts 
 - Departments 
 - Fields of study 
 - Specializations 
 - ~~Tags~~ [As per #152]
 - Companies 
   + and relationships between them

 After that, we should do migrations to fully establish database structure.

### LINKED ISSUE #167: Wrong port for MySQL Database
It seems the default exposed port for MySQL is 3306, but right now it's 5432. 
I think we should update docker config.

### LINKED ISSUE #170: Database seeding
For development purposes, application should contain factories and seeders.

- Users and companies could be created randomly
- One administrator and moderator account should always be created
- Departments, study fields and specializations could be predefined

It would be good if random company names were generated in some reasonable way.

### PR DESCRIPTION
**Pull request by sbanasiak**
We finally settled on an initial database structure. Migrations and seeders were made for each defined models. Closes #148.

Screenshot from 2022-08-23 15-45-21

For embedded models like Address, ContactDetails and Coordinates, a [laravel-castable-data-transfer-object](https://github.com/jessarcher/laravel-castable-data-transfer-object) library was used, which allows us to easily cast models to json, It also automatically verifies if proper names and types were used for fields.
This allows us to define them only with class attributes:
```php
class ContactDetails extends CastableDataTransferObject
{
    public string $email;
    public string $phone_number;
    public string $website_url;
}
```
Old generated content was removed in favor of seeders. Closes #170.

In addition, this PR closes #167.

### CODE REVIEW

File: backend/database/migrations/2022_08_11_112450_create_companies_table.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+return new class() extends Migration {
+    public function up(): void
+    {
+        Schema::create("companies", function (Blueprint $table): void {
+            $table->id();
+            $table->string("name");
+            $table->text("description");
+            $table->foreignIdFor(User::class)->constrained()->cascadeOnDelete();
+            $table->json("address");
+            $table->string("logo")->nullable();
```
Comment (mtracz) [inline backend/database/migrations/2022_08_11_112450_create_companies_table.php:20]: `logo` will be path to the file or file content encoded as base64? Maybe better use `blolb` (binary) field?
https://laravel.com/docs/9.x/migrations#column-method-binary

File: backend/database/migrations/2022_08_11_112450_create_companies_table.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+return new class() extends Migration {
+    public function up(): void
+    {
+        Schema::create("companies", function (Blueprint $table): void {
+            $table->id();
+            $table->string("name");
+            $table->text("description");
+            $table->foreignIdFor(User::class)->constrained()->cascadeOnDelete();
+            $table->json("address");
```
Comment (mtracz) [inline backend/database/migrations/2022_08_11_112450_create_companies_table.php:19]: Are you think that json field will be good? What about searching via city, and so on? The same with `contact_details` column.
I would like to keep these data as separated columns.

File: backend/database/seeders/UserSeeder.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+class UserSeeder extends Seeder
+{
+    public function run(): void
+    {
+        User::factory([
+            "role" => Role::Administrator,
+        ])->create();
+
+        User::factory(50)
+            ->has(Company::factory(rand(1, 3)))
```
Comment (mtracz) [inline backend/database/seeders/UserSeeder.php:21]: Maybe better use named arguments for such values?
```php
rand(min: 1, max: 3);
```

**Review by mtracz — CHANGES_REQUESTED** [review]

File: backend/database/migrations/2022_08_11_112450_create_companies_table.php (hunk 2 above)
Comment (krzysztofrewak) [reply backend/database/migrations/2022_08_11_112450_create_companies_table.php:19]: Laravel has options to serach in JSON fields

Comment (sbanasiak) [reply backend/database/migrations/2022_08_11_112450_create_companies_table.php:19]: > Are you think that json field will be good? What about searching via city, and so on? The same with `contact_details` column. I would like to keep these data as separated columns.

We use the laravel-castable-data-transfer-object package, which allows us to cast JSON columns to objects. In addition as above Laravel allows us to conveniently search in JSON fields: https://laravel.com/docs/9.x/queries#json-where-clauses

File: backend/database/migrations/2022_08_11_112450_create_companies_table.php (hunk 1 above)
Comment (sbanasiak) [reply backend/database/migrations/2022_08_11_112450_create_companies_table.php:20]: > `logo` will be path to the file or file content encoded as base64? Maybe better use `blolb` (binary) field? https://laravel.com/docs/9.x/migrations#column-method-binary

Support for logos will be implemented in a separate pull request. Then, most likely, the field type will also be changed.

File: backend/database/seeders/UserSeeder.php (hunk 3 above)
Comment (sbanasiak) [inline backend/database/seeders/UserSeeder.php:21]:
```suggestion
            ->has(Company::factory(rand(min: 1, max: 3)))
```

File: backend/database/migrations/2022_08_11_112450_create_companies_table.php (hunk 2 above)
Comment (mtracz) [reply backend/database/migrations/2022_08_11_112450_create_companies_table.php:19]: Ok, then I don't care. But in my opinion these fields should be extracted and put as separated columns.

**Review by mtracz — DISMISSED** [review]

File: backend/composer.json [hunk 4]
```diff
@@ -38,6 +47,27 @@
...
+      "sh -c 'cd static && php index.php --build'"
+    ],
+    "intern-populate": [
+      "sh -c 'cd static && php index.php --populate'"
+    ],
+    "intern-fetch": [
+      "sh -c 'cd static && php index.php --fetch'"
+    ],
+    "intern-purge": [
+      "rm -rf ../public/api"
```
Comment (krzysztofrewak) [inline backend/composer.json:67]: As far as i can see there is no `static` directory on your branch. Or may I missing something? https://github.com/blumilksoftware/internships/tree/148-db-preparation/backend

File: backend/app/Models/Embeddable/Address.php [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+declare(strict_types=1);
+
+namespace Internships\Models\Embeddable;
+
+use JessArcher\CastableDataTransferObject\CastableDataTransferObject;
+
+class Address extends CastableDataTransferObject
+{
+    public string $country;
```
Comment (krzysztofrewak) [inline backend/app/Models/Embeddable/Address.php:11]: Could these be readonlies?

File: .env.example [hunk 6]
```diff
@@ -12,14 +12,14 @@ LOG_LEVEL=debug
...
-DB_PORT=5432
+DB_HOST=intern-mysql
+DB_PORT=3306
 DB_DATABASE=intern
 DB_USERNAME=intern
 DB_PASSWORD=password
 DB_ROOT_PASSWORD=example
 
-DOCKER_DEV_DB_EXTERNAL_PORT=5432
+DOCKER_DEV_DB_EXTERNAL_PORT=3306
```
Comment (krzysztofrewak) [inline .env.example:22]: Please provide information in readme that you're blocking default MySQL port by default. The instruction should contain a hint how to change this external port as this probably will be a common problem.

File: backend/app/Models/Specialization.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+declare(strict_types=1);
+
+namespace Internships\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
+
+class Specialization extends Model
```
Comment (krzysztofrewak) [inline backend/app/Models/Specialization.php:12]: Please provide additional information about model's properties in docblocks. You can find an example here: https://github.com/blumilksoftware/toby/blob/main/app/Eloquent/Models/Holiday.php#L13

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]
I'm going to short vacations, so please ask @mtracz or @EwelinaLasowy to check your fixes and close my disucssions.

File: backend/composer.json (hunk 4 above)
Comment (sbanasiak) [reply backend/composer.json:67]: References to the old static backend were added by mistake while resolving merge conflicts. They are obviously unnecessary and will be removed.

**Review by EwelinaLasowy — APPROVED** [review]

**Review by mtracz — APPROVED** [review]

**Pull request merged**